src/main/java/com/mineboundteam/minebound/magic/OffensiveSpells/NecroticOffensiveSpell.java

```diff
@@ -0,0 +1,170 @@
+
+
+@Mod.EventBusSubscriber(modid = MineBound.MOD_ID)
+public class NecroticOffensiveSpell extends ActiveSpellItem {
+    public static DoubleValue FOOD_REDUCTION;
+
+    private final int manaCost;
+    private final double damageBoost;
+    private final double vampirePercent;
+
+    public final static String ACTIVE_TAG = MineBound.MOD_ID + ".necrotic_offensive.active";
+    public final static String TICK_TAG = MineBound.MOD_ID + ".necrotic_offensive.ticks";
+
+    public NecroticOffensiveSpell(Properties properties, NecroticOffensiveSpellConfig config) {
+        super(properties, config.LEVEL, MagicType.NECROTIC, SpellType.OFFENSIVE);
+        manaCost = config.MANA_COST.get();
+        damageBoost = config.DMG_BOOST.get();
+        vampirePercent = config.HEALTH_AMT.get();
+    }
+
+    @Override
+    public void use(ItemStack stack, Level level, Player player) {
+        stack.getOrCreateTag().putBoolean(ACTIVE_TAG, true);
+    }
+
+    @Override
+    public void onUsingTick(ItemStack stack, Level level, Player player) {
+        if(stack.getOrCreateTag().getBoolean(ACTIVE_TAG)) {
+            int ticks = stack.getOrCreateTag().getInt(TICK_TAG);
+            if(ticks < 20)
+                ticks++;
+            else {
+                reduceMana(manaCost, player);
+                ticks = 0; 
+            }
+            stack.getOrCreateTag().putInt(TICK_TAG, ticks);
+        }
+    }
+
+    @Override
+    public void releaseUsing(ItemStack stack, Level level, Player player) {
+        stack.getOrCreateTag().putBoolean(ACTIVE_TAG, false);
+        stack.getOrCreateTag().putInt(TICK_TAG, 0);
+    }
+
+    @SubscribeEvent
+    public static void onAttack(LivingDamageEvent event){
+        if(event.getSource().getDirectEntity() != null && event.getSource().getDirectEntity() instanceof ServerPlayer player){
+            ItemStack spell = ItemStack.EMPTY; 
+            if(player.getUseItem().getItem() instanceof NecroticOffensiveSpell)
+                spell = player.getUseItem();
+            else {
+                if(player.getCapability(PlayerSelectedSpellsProvider.PRIMARY_SPELL).isPresent()){
+                    PrimarySelected selected = player.getCapability(PlayerSelectedSpellsProvider.PRIMARY_SPELL).resolve().get(); 
+                    if(!selected.isEmpty())
+                        spell = ItemStack.of(ArmorNBTHelper.getSpellTag(player.getItemBySlot(selected.equippedSlot), ArmorNBTHelper.ACTIVE_SPELL).getCompound(selected.index));
+                }
+                if(player.getCapability(PlayerSelectedSpellsProvider.SECONDARY_SPELL).isPresent()){
+                    SecondarySelected selected = player.getCapability(PlayerSelectedSpellsProvider.SECONDARY_SPELL).resolve().get(); 
+                    if(!selected.isEmpty())
+                        spell = ItemStack.of(ArmorNBTHelper.getSpellTag(player.getItemBySlot(selected.equippedSlot), ArmorNBTHelper.ACTIVE_SPELL).getCompound(selected.index));
+                }
```

Comment:
The secondary selected spell, if one exists, will always override the primary selected spell. So if Necrotic Offensive is the primary spell, none of the code below executes since the secondary selected spell is bound to `spell`. You can take a look at the logic in the `triggerSpell` functions in `ShieldOffensiveSpell` for inspiration if you want, but you shouldn't need to do all that.

Reply:
Maybe we want to create a function in `ActiveSpellItem` for getting an active spell by type which could be used by all spells that need that functionality (Shield offensive, this, and any future spells).

Reply:
You should also use the `getSelectedSpell` function in `ActiveSpellItems` to get the spell `ItemStack`.

Reply:
Fixed; didn't know we had getSelectedSpell